### PR TITLE
Call history export import

### LINKED ISSUE #412: feature request: call log export
It would be great if there is an option to export the calls log :)

### PR DESCRIPTION
**Pull request by Merkost**
New menu items were added to the app settings.
Resolves #412

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt [hunk 1]
```diff
@@ -261,4 +290,52 @@ class SettingsActivity : SimpleActivity() {
...
+        }.onFailure {
+            toast(R.string.importing_failed)
+        }
+    }
+
+    private fun exportCallHistory(recents: ArrayList<RecentCall>, uri: Uri) {
+        if (recents.isEmpty()) {
+            toast(R.string.no_entries_for_exporting)
+        } else {
+            runCatching {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt:327]: avoid such runCaching please, we arent using it anywher and it is hard to read

Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt]: Do you mean that we use the usual try/catch instead?

Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt]: sure

File: app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt [hunk 2]
```diff
@@ -271,7 +271,7 @@ class ContactsAdapter(
     private fun setupView(view: View, contact: Contact, holder: ViewHolder) {
         view.apply {
             setupViewBackground(activity)
-            findViewById<ConstraintLayout>(R.id.item_contact_frame).isSelected = selectedKeys.contains(contact.rawId)
+            findViewById<FrameLayout>(R.id.item_contact_frame).isSelected = selectedKeys.contains(contact.rawId)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt:274]: revert this back, it is a constraintlayout

**Comment by tibbi** [comment]
add a new Migrating settings section for exporting and importing

File: app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt [hunk 3]
```diff
@@ -261,4 +290,51 @@ class SettingsActivity : SimpleActivity() {
...
+            getContent.launch(callHistoryFileType)
+        }
+
+        settings_calls_export_holder.setOnClickListener {
+            val fileName = "call_history_${getCurrentFormattedDateTime()}"
+            saveDocument.launch(fileName)
+        }
+    }
+
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/dialer/activities/SettingsActivity.kt:305]: avoid inserting double empty lines

**Comment by tibbi** [comment]
the app crashes if you import some file unrelated to phone calls

**Comment by tibbi** [comment]
add a filename inserting dialog after pressing Export, just like in the gallery favorites

**Comment by tibbi** [comment]
do not show an Exporting failed error if the user cancelled it on the file picker screen

**Comment by tibbi** [comment]
flip exporting and importing settings buttons, just like it is in our other apps

File: app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt (hunk 2 above)
Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt:274]: Screenshot 2023-07-02 at 22 15 17

How is that possible?

Comment (Merkost) [reply app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt:274]: My app crashes otherwise, because it's frameLayout in my file

**Comment by Merkost** [comment]
> flip exporting and importing settings buttons, just like it is in our other apps



In simple contacts, for example, it is import and then export, like I did here. By the way, do you think we should move it in app settings under migration settings too?

**Comment by Merkost** [comment]
> the app crashes if you import some file unrelated to phone calls

What do you mean by that? Currently, I just get an error toast when I pick an empty json file. Can you pick some other file instead of json? How to reproduce the crash?

File: app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt (hunk 2 above)
Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/dialer/adapters/ContactsAdapter.kt:274]: try updating to the latest commons

**Comment by tibbi** [comment]
what OS version are you working with?

**Comment by Merkost** [comment]
> what OS version are you working with?

Tested with api 29, 33 and 34

This what happens if the file is not associated with phone calls
Screenshot 2023-07-03 at 11 33 06

**Comment by tibbi** [comment]
try picking an invalid json then

**Comment by Merkost** [comment]
> try picking an invalid json then

I gave tried various files so far, however I still can't recreate the crash. All I got is the message in try catch that the file is invalid

File: app/src/main/res/layout/activity_settings.xml [hunk 4]
```diff
@@ -393,6 +393,48 @@
                     android:text="@string/show_incoming_calls_full_screen" />
 
             </RelativeLayout>
+
+            <include
+                android:id="@+id/settings_migration_divider"
+                layout="@layout/divider" />
+
+            <TextView
+                android:id="@+id/settings_migration_section_label"
```
Comment (tibbi) [inline app/src/main/res/layout/activity_settings.xml:402]: the label doesnt change its color properly, add it in onResumes list of section titles

**Comment by Merkost** [comment]
I have added Kotlin serialisation instead of using Json for deserialising as that wan not working properly. Could you test it now please.

**Comment by tibbi** [comment]
right, seems to work just fine now. One more detail, can you change the file extension to .txt? Just like we export favorites and settings in the gallery. Json isnt recognized as a text file so if you click on it with a file manager, by default no text editors show up.

**Comment by Merkost** [comment]
> right, seems to work just fine now. One more detail, can you change the file extension to .txt? Just like we export favorites and settings in the gallery. Json isnt recognized as a text file so if you click on it with a file manager, by default no text editors show up.

We are exporting json file, why do we need to allow picking any other formats like txt?

**Comment by tibbi** [comment]
just so that it is easier to open by the user. Wont create any complications on our side.

**Comment by Merkost** [comment]
> just so that it is easier to open by the user. Wont create any complications on our side.

so you want me to change exporting and importing formats to txt, right? No jsons anymore

**Comment by tibbi** [comment]
actually nevermind, lets keep it as json. Settings are exported some other way, it isnt a json at all

**Comment by tibbi** [comment]
everytime you work with jsons you should test the app in release mode too, now it gives an error at exporting. It needs some proguard rules. And split the `setupCallsImportExport` settings function, handle import and export separately.

**Comment by tibbi** [comment]
alright thanks, seems to work just fine

**Pull request merged**